NR-305349 - NRMAAnalytics lazy initialization

smalsam-newr opened the pull request:
Check if NRMAAnalytics is already initialized before blindly re-initializing it again.

This still needs to be tested for background instrumentation. 

smalsam-newr committed:
NR-305349 - NRMAAnalytics lazy initialization

Check if NRMAAnalytics is already initialized before blindly re-initializing it again.

cdillard-NewRelic reviewed (dismissed):
👍 tested with BG reporting and other than a double harvest on bg/fg that was preexisting it looks good. Approved.

mbruin-NR reviewed (commented):
There is a bug when background mode is enabled. Because the analyticsController never get's destroyed in background mode with this change the analyticsController will never get recreated. Because of this the _sessionStartTime never changes on a new session.
The 'appSessionStartDate' get's updated in 'sessionStartInitialization' and then as part of 'initializeAnalytics' that value is passed into the newly initialized 'analyticsController'. Make sure the _sessionStartTime in the analyticsController reflects a new session.

smalsam-newr committed:
Start a new session after background

When coming out of the background, we want to be able to start a new session, with a new session start time. This enables that.

mbruin-NR commented on Agent/General/NewRelicAgentInternal.m:
```diff
@@ -445,7 +445,15 @@ - (void) memoryUsageDidUpdate:(NSNotification*)notification {
+        if(!self.analyticsController) {
+            self.analyticsController = [[NRMAAnalytics alloc] initWithSessionStartTimeMS:(long long)([self.appSessionStartDate timeIntervalSince1970] * 1000)];
+        } else if(didFireEnterForeground && didFireEnterBackground) {
+            // We are coming back to the foreground after having a background stint
```
There are some comments that can be removed here.
